### PR TITLE
Fix a long-standing bug where the user directory would return 1 more row than requested.

### PR DESCRIPTION
**Pull request by reivilibre**
Fixes: #10840 

This is a very quick drive-by fix because I noticed someone running into it in a client dev room.

This pull request is commit-by-commit review friendly. 

Original commit schedule, with full messages:

<ol>
<li>

Respect the limit in the return of the user directory 

</li>
<li>

Add a test to ensure this doesn't happen again 

</li>
<li>

Clamp `limit` because it would be janky for negative numbers to slice backwards in the array 

</li>
</ol>

### CODE REVIEW

**Comment by clokep** [comment]
Oh this was still a draft, sorry. :(

**Comment by clokep** [comment]
Maybe fixes #10840?

**Comment by reivilibre** [comment]
I won't turn down the review. Draft is just to let CI run before I bother anyone with it (in theory) :-)

File: synapse/rest/client/user_directory.py [hunk 1]
```diff
@@ -63,8 +63,8 @@ async def on_POST(self, request: SynapseRequest) -> Tuple[int, JsonMapping]:
 
         body = parse_json_object_from_request(request)
 
-        limit = body.get("limit", 10)
-        limit = min(limit, 50)
+        limit = int(body.get("limit", 10))
+        limit = max(min(limit, 50), 0)
```
Comment (DMRobertson) [inline synapse/rest/client/user_directory.py:67]: The call to `int` sneakily allows the limit to be given as a string. I can look the other way for a drive-by fix though.

**Review by DMRobertson — APPROVED** [review]

**Pull request merged**